_posts/2025-03-27-Boost-OpenSearch-VectorSearch-Performance-With-Intel-AVX512.md

```diff
@@ -0,0 +1,187 @@
+* **SQFP16 Encoding:**  
+  * Hot function: *faiss::query\_to\_code*.  
+  * Benefits of AVX-512:  
+    * Indexing: Up to 17% reduction in cycles.  
+    * Search: Up to 6% reduction in cycles.
+
+| Inner Product space type |  |  |  |  |
+| ----- | :---- | :---- | :---- | :---- |
+|  | Encoding | Function | %Cycle Spent (AVX2) | %Cycle Spent (AVX512) |
+| Indexing | fp32 | fvec\_inner\_product | 28.86 | 7.32 |
+| Indexing | SQfp16 | query\_to\_code | 17.95 | 10.94 |
+| Search | fp32 | fvec\_inner\_product\_batch\_4 | 34.66 | 31.74 |
+| Search | SQfp16 | query\_to\_code | 42.24 | 37.73 |
+
+| L2 space type |  |  |  |  |
+| ----- | :---- | :---- | :---- | :---- |
+|  | Encoding | Function | %Cycle Spent (AVX2) | %Cycle Spent (AVX512) |
+| Indexing | fp32 | fvec\_L2sqr | 36.76 | 16.75 |
+| Indexing | SQfp16 | query\_to\_code | 26.18 | 21.61 |
+| Search | fp32 | fvec\_L2sqr\_bactch\_4 | 31.80 | 28.32 |
+| Search | SQfp16 | query\_to\_code | 36.99 | 34.72 |
+
+Starting with OpenSearch version 2.18, AVX-512 is enabled by default. As of March 2025, OpenSearch has shown best performance for AVX-512 on AWS r7i instances. 
+
+The next section describes benchmark results that were run with AVX2 and AVX-512 versions of the Faiss library shipped for x64 architecture ([Documentation](https://opensearch.org/docs/latest/search-plugins/knn/knn-index/#supported-lucene-methods:~:text=multiple%20of%208.-,x64%20architecture,libopensearchknn_faiss.so%3A%20The%20non%2Doptimized%20Faiss%20library%20without%20SIMD%20instructions.,-When%20using%20the)). These benchmarks were run using [OpenSearch Benchmark (OSB) vector search workload](https://github.com/opensearch-project/opensearch-benchmark-workloads/tree/main/vectorsearch) using the following [Benchmark Configuration](https://github.com/opensearch-project/project-website/issues/3697#issuecomment-2771024897)
+
+## **Results**
+
+The results show that the time spent in hot functions of the distance calculation is significantly reduced due to AVX-512, and the OpenSearch cluster shows higher throughput on search and indexing. 
+
+SQfp16 encoding provided by the Faiss library further helps with faster computation and efficient storage by compressing the 32-bit floating point vectors into 16-bit floating point format. The smaller memory footprints allows for more vectors to be stored in the same amount of memory while the operations on the 16-bits floats are typically faster than those on 32-bit floats, leading to quicker similarity searches. 
+
+A higher performance improvement is observed between AVX512 and AVX2 on fp16, due to code optimizations and the use of AVX-512 intrinsics in Faiss, which are not present on AVX2. 
+
+A general observation across all benchmarks is that AVX-512 improves performance due to significant reduction in [pathlength](https://en.wikipedia.org/wiki/Instruction_path_length), which is the number of machine instructions needed to execute a workload. 
```

Comment:
```suggestion
A general observation across all benchmarks is that AVX-512 improves performance because of a significant reduction in [path length](https://en.wikipedia.org/wiki/Instruction_path_length)---the number of machine instructions needed to execute a workload. 
```